### DISCUSSION TITLE
How to reference entire context as an object?

### DISCUSSION THREAD

**Discussion by wez** [mitsuhiko/minijinja#344] [Q&A] [answered]
In my application I have a struct that is `Serialize` being passed in as my context.
I would like to allow my users to optionally do something like `{{ context | tojson }}`.

Is there built-in syntax or a built-in function that can provide a reference to the overall context like this?
If not, what's the recommended way to approach this?

**Comment by cataggar** [comment]
May be you are looking for [add_global](https://docs.rs/minijinja/latest/minijinja/struct.Environment.html#method.add_global). [Here](https://github.com/cataggar/configur/blob/fe3207bdf1f5333b0cfc19687a21fe56a98fd345/src/jinga.rs#L118) is how I used it.

**Comment by mitsuhiko** [comment, marked as answer]
There is some risk with this behavior because it makes that you have a self-referential data structure if you are not careful. In pure theory it would be possible to write a `globals()` function (if an API was added to enumerate all variables in MiniJinja) that returns all variables, but even then you end up with the situation that actual globals are included that you probably don't want such as `range()`, `debug()` or that hypothetical `globals()` function.

What I would do instead is to say that you wrap the context that you pass for rendering with a helper that makes it self-referential within reason (eg: `CONTEXT.name` is possible, but `CONTEXT.CONTEXT.name` would fail). I added a simple example here that shows how this goes: https://github.com/mitsuhiko/minijinja/blob/main/examples/self-referential-context/src/main.rs

In short, you would need to change `template.render(context!( ... ))` with `template.render(make_self_referential(context!( ... )))`.
